### DISCUSSION TITLE
why 2 ways of using state?

### DISCUSSION THREAD

**Discussion by ghost** [LegendApp/legend-state#155] [Q&A]

**Comment by jmeistrich** [comment]
Two main reasons:

1. The fine-grained techniques can get messy in complex components that run a lot of code to determine what/how to render based on lots of options, like a Button component. In those cases it can be better to have the render function do its normal React thing rather than have tons of computed properties.
2. It's better for gradual adoption. For teams that are used to React or migrating from a different state library, it's easier to adopt and migrate when its more familiar. Then as they get used to it, they can start adopting more of the fine-grained techniques where its useful. If they had to jump straight into only fine-grained it would be a blocker.

For what it's worth, we write most of our new code with the fine-grained components, though complex UI components with lots of options are easier to do in the normal React way with just `.use()` on the observable props. And we still have old components that are not fine-grained but it's not worth the time to change.

If you want to go all in on fine-grained techniques that's great! I find it a lot more fun and easy to reason about personally 🤟
